File: backend/app/schemas/plaza.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, field_serializer

from app.core.time import to_app_timezone


class PlazaAuthor(BaseModel):
    session_id: str
    nickname: str
    avatar_url: Optional[str] = None
    ebti_type: Optional[str] = None
    ebti_name: Optional[str] = None


class PlazaPostItem(BaseModel):
    id: int
    title: Optional[str] = None
    content: str
    mood_label: Optional[str] = None
    location: Optional[str] = None
    tags: list[str] = []
    media_url: Optional[str] = None
    author: PlazaAuthor
    like_count: int = 0
    comment_count: int = 0
    repost_count: int = 0
    liked: bool = False
    xiaoxi_liked: bool = False
    reposted: bool = False
    published_at: Optional[datetime] = None
    created_at: datetime

    @field_serializer("published_at", "created_at")
    def serialize_time(self, value: Optional[datetime]) -> Optional[str]:
        if value is None:
            return None
        return to_app_timezone(value).isoformat()


class PlazaCommentCreate(BaseModel):
    session_id: str = Field(..., min_length=1, max_length=80)
    content: str = Field(..., min_length=1, max_length=800)


class PlazaCommentReplyItem(BaseModel):
    id: int
    record_id: int
    parent_id: int
    reply_to_comment_id: Optional[int] = None
    reply_to_author: Optional[PlazaAuthor] = None
    author: PlazaAuthor
    author_type: str
    content: str
    like_count: int = 0
    created_at: datetime

    @field_serializer("created_at")
    def serialize_created_at(self, value: datetime) -> str:
        return to_app_timezone(value).isoformat()


class PlazaCommentItem(BaseModel):
    id: int
    record_id: int
    parent_id: Optional[int] = None
    reply_to_comment_id: Optional[int] = None
    reply_to_author: Optional[PlazaAuthor] = None
    author: PlazaAuthor
    author_type: str
    content: str
    like_count: int = 0
    reply_count: int = 0
    replies: list[PlazaCommentReplyItem] = Field(default_factory=list)
    created_at: datetime

    @field_serializer("created_at")
    def serialize_created_at(self, value: datetime) -> str:
        return to_app_timezone(value).isoformat()

File: frontend_update/frontend/src/api/plaza.ts
import api from './client'

export interface PlazaAuthor {
  session_id: string
  nickname: string
  avatar_url?: string | null
  ebti_type?: string | null
  ebti_name?: string | null
}

export interface PlazaPostItem {
  id: number
  title?: string | null
  content: string
  mood_label?: string | null
  location?: string | null
  tags: string[]
  media_url?: string | null
  author: PlazaAuthor
  like_count: number
  comment_count: number
  repost_count: number
  liked: boolean
  xiaoxi_liked: boolean
  reposted: boolean
  published_at?: string | null
  created_at: string
}

export interface PlazaCommentItem {
  id: number
  record_id: number
  parent_id?: number | null
  reply_to_comment_id?: number | null
  reply_to_author?: PlazaAuthor | null
  author: PlazaAuthor
  author_type: 'user' | 'xiaoxi'
  content: string
  like_count: number
  reply_count: number
  replies: PlazaCommentReplyItem[]
  created_at: string
}

export interface PlazaCommentReplyItem {
  id: number
  record_id: number
  parent_id: number
  reply_to_comment_id?: number | null
  reply_to_author?: PlazaAuthor | null
  author: PlazaAuthor
  author_type: 'user' | 'xiaoxi'
  content: string
  like_count: number
  created_at: string
}

export const getPlazaPosts = (sid: string) =>
  api.get<PlazaPostItem[]>(`/plaza/posts?session_id=${encodeURIComponent(sid)}`)

export const likePlazaPost = (id: number, sid: string) =>
  api.post<PlazaPostItem>(`/plaza/posts/${id}/like?session_id=${encodeURIComponent(sid)}`)

export const unlikePlazaPost = (id: number, sid: string) =>
  api.delete<PlazaPostItem>(`/plaza/posts/${id}/like?session_id=${encodeURIComponent(sid)}`)

export const repostPlazaPost = (id: number, sid: string) =>
  api.post<PlazaPostItem>(`/plaza/posts/${id}/repost?session_id=${encodeURIComponent(sid)}`)

export const getPlazaComments = (id: number) =>
  api.get<PlazaCommentItem[]>(`/plaza/posts/${id}/comments`)

export const createPlazaComment = (id: number, sid: string, content: string) =>
  api.post<PlazaCommentItem>(`/plaza/posts/${id}/comments`, {
    session_id: sid,
    content,
  })

export const getPlazaCommentReplies = (id: number) =>
  api.get<PlazaCommentReplyItem[]>(`/plaza/comments/${id}/replies`)

export const createPlazaCommentReply = (id: number, sid: string, content: string) =>
  api.post<PlazaCommentReplyItem>(`/plaza/comments/${id}/replies`, {
    session_id: sid,
    content,
  })
